### PR TITLE
feat: [Session Replay] Add `UIHostingView` recorder

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Record SwiftUI Hosting view using [custom reflection](https://github.com/DataDog/dd-sdk-ios/blob/maxep/sr/swift-ui/DatadogSessionReplay/Sources/Utilities/ReflectionMirror.swift) to inspect the displayed SwiftUI items.

Follow up: https://github.com/DataDog/dd-sdk-ios/pull/1907

### Note

This implementation is still lacking good test coverage and clean-up. SwiftUI support in Session Replay is an on-going effort which needs merge to `develop` for peer contribution and dogfooding.

### Reflection Mirror

A representation of the substructure and display style of an instance of any type.

The `ReflectionMirror` defers from the standard ``Mirror`` in some key aspect of its implementation:

- **Display style**
The display styles are based on the type metadata and does not define additional `optional`, `collection`, `dictionary`, or `set` which are set by ``CustomReflectable`` implementation is standard mirroring.
Additionally, the `.enum` style will include the case name as associated value.

- **Optional type**
If the subject type is optional, the display style will either be `nil` or represent the wrapped subject type. If the subject is not `nil`, reflection will be applied to the wrapped value instead of the subject itself.

- **Lazy inspection**
All inspection endpoints are lazily loading child value, even when accessing property by key (this is not the case with the standard `Mirror`). This is made possible by only loading the fields metadata but not their values.
The `superclassMirror` is also lazily loaded, again this is not the case with the standard `Mirror`.

- **Ignore custom reflectable**
The `ReflectionMirror` ignores the ``CustomReflectable`` or ``CustomLeafReflectable`` protocols and treat any conforming objects as `Any.Type`.

### SwiftUI Tree

A SwiftUI tree is called a `DisplayList` where a node item is defined as following:
```swift
/// An item is an element of the displayed tree in `SwiftUI`.
///
/// The item has a unique identity, a frame in the current branch, and a value.
/// A value can be:
/// - An effect (node): which applied an effect on a sub-tree.
/// - A content (leaf): The content of the displayed item.
internal struct Item {
  internal enum Value {
    case effect(Effect, DisplayList)
    case content(Content)
    case unknown
  }

  let identity: Identity
  let frame: CGRect
  let value: Value
}
```

A displayed item can be rendered in a subview that is stored in a view cache and linked using the `identity` property. Both  the `DisplayList` and the `ViewCache` are reflecting to build the wireframes:
- The `ViewCache` is reflected on the main thread to access the views and layers properties
- The `DisplayList` is reflected lazily on the background thread for performances.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes
- [ ] Add Objective-C interface for public APIs (see our [guidelines](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3157787243/RFC+-+Modular+Objective-C+Interface#Recommended-solution) [internal]) and run `make api-surface`)

### CODE REVIEW

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        case .unknown:
+            return nil // Need a placeholder
+        }
+    }
+
+    private func contentWireframe(item: DisplayList.Item, content: DisplayList.Content, context: Context) -> [SRWireframe] {
+        contentWireframe(item: item, content: content, context: context).map { [$0] } ?? []
+    }
+
+    func effectWireframe(item: DisplayList.Item, effect: DisplayList.Effect, list: DisplayList, context: Context) -> [SRWireframe] {
```
Comment (mariedm) [inline DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift:120]: nit
```suggestion
    private func effectWireframe(item: DisplayList.Item, effect: DisplayList.Effect, list: DisplayList, context: Context) -> [SRWireframe] {
```

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                borderColor: viewInfo?.borderColor,
+                borderWidth: viewInfo?.borderWidth,
+                backgroundColor: viewInfo?.backgroundColor,
+                cornerRadius: viewInfo?.cornerRadius,
+                opacity: viewInfo?.alpha
+            )
+        case .platformView:
+            return nil // Should be recorder by UIKit recorder
+        case .unknown:
+            return nil // Need a placeholder
```
Comment (mariedm) [inline DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift:112]: Should we return an "unsupported" placeholder wireframe then?

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            return buildWireframes(items: list.items, context: context)
+
+        case .identify, .unknown:
+            return buildWireframes(items: list.items, context: context)
+        }
+    }
+}
+
+@available(iOS 13.0, *)
+extension SwiftUIWireframesBuilder.Context {
```
Comment (mariedm) [inline DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift:140]: nit
```suggestion
internal extension SwiftUIWireframesBuilder.Context {
```

File: DatadogSessionReplay/Tests/Utilities/ReflectionMirrorTests.swift [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        XCTAssertEqual(mirror.displayStyle, .nil)
+    }
+
+    func testNonNilDisplay() {
+        struct Mock {}
+        let mirror = ReflectionMirror(reflecting: Optional.some(Mock()) as Any)
+        XCTAssertEqual(mirror.displayStyle, .struct)
+    }
+
+//    func testExistentialStructDisplay() {
```
Comment (mariedm) [inline DatadogSessionReplay/Tests/Utilities/ReflectionMirrorTests.swift:51]: Is this a leftover?

File: IntegrationTests/IntegrationScenarios/Scenarios/SessionReplay/SRMultipleViewsRecordingScenarioTests.swift [hunk 5]
```diff
@@ -25,10 +25,10 @@ class SRMultipleViewsRecordingScenarioTests: IntegrationTests, RUMCommonAsserts,
     /// will highly depend on the performance of app, Simulator and CI.
     private struct Baseline {
         /// Number of all SR segments to pass the test.
-        static let totalSegmentsCount = 15
+        static let totalSegmentsCount = 14
         /// Number of all SR records to pass the test.
         static let totalRecordsCount = 35
-        /// Number of all "full snapshot" records to pass the test.
+        /// Number of all "full snapshot" records to pasOs the test.
```
Comment (mariedm) [inline IntegrationTests/IntegrationScenarios/Scenarios/SessionReplay/SRMultipleViewsRecordingScenarioTests.swift:31]: nit
```suggestion
        /// Number of all "full snapshot" records to pass the test.
```

File: DatadogSessionReplay/Tests/Utilities/ReflectionMirrorTests.swift (hunk 4 above)
Comment (maxep) [reply DatadogSessionReplay/Tests/Utilities/ReflectionMirrorTests.swift:51]: Yes, I will remove this.

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift (hunk 2 above)
Comment (maxep) [reply DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift:112]: Yes, lets add this in further PR

**Review by mariedm — APPROVED** [review]
Awesome work! 🙌 

Not much to add from my previous review in your feature branch, apart from unit tests, but we'll tackle this in a separate PR. LGMT! 🚀

**Review by mariedm — APPROVED** [review]

**Pull request merged**